File: calculate_path_length.py
#!/usr/bin/env python3
"""
Calculate the total length of a MultiLineString geometry.
Uses Haversine formula for great circle distances and accounts for elevation changes.
"""

import math
import json

def haversine_distance(lat1, lon1, lat2, lon2):
    """
    Calculate the great circle distance between two points on Earth.
    Returns distance in meters.
    """
    R = 6371000  # Earth's radius in meters
    
    lat1_rad = math.radians(lat1)
    lat2_rad = math.radians(lat2)
    delta_lat = math.radians(lat2 - lat1)
    delta_lon = math.radians(lon2 - lon1)
    
    a = (math.sin(delta_lat / 2) ** 2 + 
         math.cos(lat1_rad) * math.cos(lat2_rad) * math.sin(delta_lon / 2) ** 2)
    c = 2 * math.atan2(math.sqrt(a), math.sqrt(1 - a))
    
    return R * c

def calculate_3d_distance(lat1, lon1, elev1, lat2, lon2, elev2):
    """
    Calculate 3D distance between two points, accounting for elevation.
    Returns distance in meters.
    """
    # Calculate horizontal distance using Haversine
    horizontal_dist = haversine_distance(lat1, lon1, lat2, lon2)
    
    # Calculate elevation difference
    elev_diff = abs(elev2 - elev1)
    
    # Calculate 3D distance using Pythagorean theorem
    distance_3d = math.sqrt(horizontal_dist**2 + elev_diff**2)
    
    return distance_3d

def calculate_path_length(geometry):
    """
    Calculate total length of a MultiLineString geometry.
    """
    total_length = 0.0
    
    for line_string in geometry['coordinates']:
        for i in range(len(line_string) - 1):
            point1 = line_string[i]
            point2 = line_string[i + 1]
            
            lon1, lat1, elev1 = point1
            lon2, lat2, elev2 = point2
            
            segment_length = calculate_3d_distance(lat1, lon1, elev1, lat2, lon2, elev2)
            total_length += segment_length
    
    return total_length

# The MultiLineString geometry from the user
geometry = {
    "type": "MultiLineString",
    "coordinates": [
        [
            [-105.29199, 39.99753, 0],
            [-105.292067643, 39.997489607, 1776.32800293],
            [-105.292196195, 39.997417204, 1775.205078125],
            [-105.292289874, 39.997407958, 1773.150634766],
            [-105.292359969, 39.997362733, 1769.647216797],
            [-105.292395, 39.99735, 1769.647216797]
        ],
        [
            [-105.283755, 39.97359, 0],
            [-105.284027296, 39.973406519, 1855.171020508],
            [-105.284132974, 39.973284298, 1856.256103516],
            [-105.284279741, 39.972887249, 1858.844238281],
            [-105.284283026, 39.972805889, 1856.794189453],
            [-105.284219921, 39.972770597, 1855.615112305],
            [-105.284076881, 39.972754525, 1855.2578125],
            [-105.283751462, 39.972842897, 1859.276367188],
            [-105.283583329, 39.97286382, 1862.946899414],
            [-105.283435269, 39.972789924, 1863.522094727],
            [-105.283336198, 39.972699445, 1865.359741211],
            [-105.283044038, 39.972298811, 1873.10168457],
            [-105.282806693, 39.972039484, 1876.188720703],
            [-105.282734216, 39.971777674, 1879.998535156],
            [-105.282497515, 39.971652749, 1883.660522461],
            [-105.282453116, 39.971600771, 1883.660522461],
            [-105.282443679, 39.971539956, 1885.229248047],
            [-105.282519643, 39.970804899, 1897.714599609],
            [-105.282577055, 39.970646148, 1900.794799805],
            [-105.282630778, 39.970607061, 1900.794799805],
            [-105.282724414, 39.970616261, 1899.158569336],
            [-105.282996339, 39.970746641, 1900.392456055],
            [-105.283186829, 39.970880019, 1902.895751953],
            [-105.28344, 39.97107, 1906.433837891]
        ]
        # Note: The full geometry has many more line segments, but I'm including
        # just the first two for demonstration. The full calculation would include all segments.
    ]
}

if __name__ == "__main__":
    # For demonstration, let's calculate with just the first two line strings
    # In practice, you'd include all the coordinates from your geometry
    
    total_length = calculate_path_length(geometry)
    
    print(f"Total path length: {total_length:.2f} meters")
    print(f"Total path length: {total_length/1000:.2f} kilometers")
    print(f"Total path length: {total_length*3.28084:.2f} feet")
    print(f"Total path length: {total_length*3.28084/5280:.2f} miles")

File: calculate-path-length.js
const fs = require('fs');

// Haversine formula for calculating distance between two points on Earth
function haversineDistance(lat1, lon1, lat2, lon2) {
    const R = 6371; // Earth's radius in kilometers
    const dLat = (lat2 - lat1) * Math.PI / 180;
    const dLon = (lon2 - lon1) * Math.PI / 180;
    const a = Math.sin(dLat/2) * Math.sin(dLat/2) +
              Math.cos(lat1 * Math.PI / 180) * Math.cos(lat2 * Math.PI / 180) *
              Math.sin(dLon/2) * Math.sin(dLon/2);
    const c = 2 * Math.atan2(Math.sqrt(a), Math.sqrt(1-a));
    return R * c;
}

// Calculate 3D distance including elevation
function calculate3DDistance(lat1, lon1, elev1, lat2, lon2, elev2) {
    const horizontalDistance = haversineDistance(lat1, lon1, lat2, lon2);
    const verticalDistance = Math.abs(elev2 - elev1) / 1000; // Convert meters to kilometers
    return Math.sqrt(horizontalDistance * horizontalDistance + verticalDistance * verticalDistance);
}

// Read and parse the GeoJSON file
const geojsonData = JSON.parse(fs.readFileSync('test-output/tmp.geojson', 'utf8'));

let totalLength = 0;
let segmentCount = 0;
let pointCount = 0;

console.log('🔍 Analyzing path geometry...');

if (geojsonData.geometry && geojsonData.geometry.type === 'MultiLineString') {
    const coordinates = geojsonData.geometry.coordinates;
    
    coordinates.forEach((lineString, lineIndex) => {
        console.log(`   📍 Processing LineString ${lineIndex + 1} with ${lineString.length} points`);
        
        for (let i = 0; i < lineString.length - 1; i++) {
            const [lon1, lat1, elev1] = lineString[i];
            const [lon2, lat2, elev2] = lineString[i + 1];
            
            const segmentLength = calculate3DDistance(lat1, lon1, elev1, lat2, lon2, elev2);
            totalLength += segmentLength;
            segmentCount++;
            pointCount++;
        }
        pointCount++; // Count the last point of each line
    });
    
    console.log('\n📊 PATH ANALYSIS RESULTS:');
    console.log(`   • Total segments: ${segmentCount}`);
    console.log(`   • Total points: ${pointCount}`);
    console.log(`   • Total length: ${totalLength.toFixed(6)} km`);
    console.log(`   • Total length: ${(totalLength * 1000).toFixed(2)} meters`);
    console.log(`   • Total length: ${(totalLength * 0.621371).toFixed(6)} miles`);
    
    // Calculate elevation statistics
    let allElevations = [];
    coordinates.forEach(lineString => {
        lineString.forEach(([lon, lat, elev]) => {
            allElevations.push(elev);
        });
    });
    
    const minElev = Math.min(...allElevations);
    const maxElev = Math.max(...allElevations);
    const avgElev = allElevations.reduce((a, b) => a + b, 0) / allElevations.length;
    
    console.log(`\n🏔️  ELEVATION STATISTICS:`);
    console.log(`   • Minimum elevation: ${minElev.toFixed(2)} meters`);
    console.log(`   • Maximum elevation: ${maxElev.toFixed(2)} meters`);
    console.log(`   • Elevation gain: ${(maxElev - minElev).toFixed(2)} meters`);
    console.log(`   • Average elevation: ${avgElev.toFixed(2)} meters`);
    
} else {
    console.log('❌ Error: Expected MultiLineString geometry');
    console.log(`   Found: ${geojsonData.geometry ? geojsonData.geometry.type : 'No geometry'}`);
}
